README.md
# appcrawler

Appcrawler是一个基于自动遍历的App爬虫工具，支持Android和IOS，支持真机和模拟器。最大的特点是灵活性高，可通过配置来设定遍历的规则。

## quick start

```bash
-------------------------------------------------
appcrawler 全平台自动遍历测试工具
Q&A: https://ceshiren.com/c/opensource/appcrawler
author: seveniruby
-------------------------------------------------


Usage: appcrawler [options]

  -a, --app <value>        Android或者iOS的文件地址, 可以是网络地址, 赋值给appium的app选项
  -e, --encoding <value>   set encoding, such as UTF-8 GBK
  -c, --conf <value>       配置文件地址
  -p, --platform <value>   平台类型android或者ios, 默认会根据app后缀名自动判断
  -t, --maxTime <value>    最大运行时间. 单位为秒. 超过此值会退出. 默认最长运行3个小时
  -u, --appium <value>     appium的url地址
  -o, --output <value>     遍历结果的保存目录. 里面会存放遍历生成的截图, 思维导图和日志
  --capability k1=v1,k2=v2...
                           appium capability选项, 这个参数会覆盖-c指定的配置模板参数, 用于在模板配置之上的参数微调
  -y, --yaml <value>       代表配置的yaml语法，比如blackList: [ {xpath: action_night } ]，用于避免使用配置文件的情况
  -r, --report <value>     输出html和xml报告
  --template <value>       输出代码模板
  --master <value>         master的diff.yml文件地址
  --candidate <value>      candidate环境的diff.yml文件
  -v, --verbose-debug      是否展示更多debug信息
  -vv, --verbose-trace     是否展示更多trace信息
  --demo                   生成demo配置文件学习使用方法
  --help
                           示例
                           appcrawler -a xueqiu.apk
                           appcrawler -a xueqiu.apk --capability noReset=true
                           appcrawler -c conf/xueqiu.json -p android -o result/
                           appcrawler -c xueqiu.yaml --capability udid=[你的udid] -a Snowball.app
                           appcrawler -c xueqiu.yaml -a Snowball.app -u 4730
                           appcrawler -c xueqiu.yaml -a Snowball.app -u http://127.0.0.1:4730/wd/hub

                           #生成demo配置文件到当前目录下的demo.yaml
                           appcrawler --demo

                           #启动已经安装过的app
                           appcrawler --capability "appPackage=com.xueqiu.android,appActivity=.view.WelcomeActivityAlias"

                           #使用yaml参数
                           appcrawler -a xueqiu.apk -y "blackList: [ {xpath: action_night}, {xpath: '.*[0-9\\.]{2}.*'} ]"

                           #从已经结束的结果中重新生成报告
                           appcrawler --report result/

```

## 配置文件格式

### 执行参数与配置文件

- capability设置：与appium完全一致
- testcase：用于启动app后的基础测试用例
- selectedList：遍历范围设定
- triggerActions：特定条件触发执行动作的设置
- selectedList：需要被遍历的元素范围
- firstList：优先被点击
- lastList：最后被点击
- tagLimitMax：同祖先（同类型）的元素最多点击多少次
- backButton：当所有元素都被点击后默认后退控件定位
- blackList：黑名单
- maxDepth: 遍历的最大深度

### 配置的最小单元 测试用例模型

testcase的完整形态

- given：所有的先决条件
- when：先决条件成立后的行为
- then：断言集合

testcase的简写形态

- xpath：对应when里的xpath
- action：对应when的action

执行参数比配置文件优先级别高

- given 前提条件
- when 执行动作
- then 写断言

简写形态

- xpath xpath支持xpath表达式、正则、包含
- action 支持

### xpath定义

- xpath
    - //*[@resource-id=‘xxxx’]
    - //*[contains(@text, ‘密码’)]
- 正则
    - ^确定$
    - ^.*输入密码
- 包含
    - 密码
    - 输入
    - 请

### action定义

- "" 只是截图记录
- back 后退
- backApp 回退到当前的app 默认等价于back行为 可定制
- monkey 随机事件
- click
- longTap
- xxx() 执行scala或者java代码
    - Thread.sleep(3000)
    - driver.swipe(0.9, 0.5, 0.1, 0.5)
- 非以上所有行为是输入 xx ddd

### 完整配置文件

```yaml
---
maxTimeDescription: "最大运行时间"
maxTime: 10800
maxDepthDescription: "默认的最大深度10, 结合baseUrl可很好的控制遍历的范围"
maxDepth: 10
capabilityDescription: "appium的capability通用配置，其中automationName代表自动化的驱动引擎，除了支持appium的\
  所有引擎外，额外增加了adb和selenium的支持"
capability:
  appActivity: ".ApiDemos"
  appium: "http://127.0.0.1:4723/wd/hub"
  noReset: "true"
  appPackage: "io.appium.android.apis"
  fullReset: "false"
testcaseDescription: "测试用例设置，用于遍历开始之前的一些前置操作，比如自动登录"
testcase:
  name: "AppCrawler TestCase"
  steps:
    - given: []
      when: null
      then: []
      xpath: "/*/*"
      action: "Thread.sleep(1000)"
      actions: []
      times: -1
triggerActionsDescription: "在遍历过程中需要随时处理的一些操作，比如弹框、登录等"
triggerActions:
  - given: []
    when: null
    then: []
    xpath: "permission_allow_button"
    action: ""
    actions: []
    times: 3
  - given: []
    when: null
    then: []
    xpath: "允许"
    action: ""
    actions: []
    times: 3
selectedListDescription: "默认遍历列表，只有出现在这个列表里的控件范围才会被遍历"
selectedList:
  - given: []
    when: null
    then: []
    xpath: "//*[contains(name(), 'Button')]"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[contains(name(), 'Text') and @clickable='true' and string-length(@text)<10]"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[@clickable='true']//*[contains(name(), 'Text') and string-length(@text)<10]"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[contains(name(), 'Image') and @clickable='true']"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[@clickable='true']/*[contains(name(), 'Image')]"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[contains(name(), 'Image') and @name!='']"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[contains(name(), 'Text') and @name!='' and string-length(@label)<10]"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//a"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[contains(@class, 'Text') and @clickable='true' and string-length(@text)<10]"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[@clickable='true']//*[contains(@class, 'Text') and string-length(@text)<10]"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[contains(@class, 'Image') and @clickable='true']"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[@clickable='true']/*[contains(@class, 'Image')]"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[@clickable='true' and contains(@class, 'Button')]"
    action: ""
    actions: []
    times: -1
blackListDescription: "黑名单列表 matches风格, 默认排除内容包含2个数字的控件"
blackList:
  - given: []
    when: null
    then: []
    xpath: ".*[0-9]{2}.*"
    action: ""
    actions: []
    times: -1
firstListDescription: "优先遍历列表，同时出现在selectedList与firstList中的控件会被优先遍历"
firstList: []
lastListDescription: "最后遍历列表，同时出现在selectedList与lastList中的控件会被最后遍历"
lastList:
  - given: []
    when: null
    then: []
    xpath: "//*[@selected='true']/..//*"
    action: ""
    actions: []
    times: -1
  - given: []
    when: null
    then: []
    xpath: "//*[@selected='true']/../..//*"
    action: ""
    actions: []
    times: -1
backButtonDescription: "后退按钮列表，默认在所有控件遍历完成后，才会最后点击后退按钮。目前具备了自动判断返回按钮的能力，默认不需要配置"
backButton:
  - given: []
    when: null
    then: []
    xpath: "Navigate up"
    action: ""
    actions: []
    times: -1
xpathAttributesDescription: "在生成一个控件的唯一定位符中应该包含的关键属性"
xpathAttributes:
  - "name()"
  - "name"
  - "label"
  - "value"
  - "resource-id"
  - "content-desc"
  - "text"
  - "id"
  - "name"
  - "innerText"
  - "tag"
  - "class"
sortByAttributeDescription: "陆续根据属性进行遍历排序微调，depth表示从dom中最深层的控件开始遍历，list表示dom中列表优先，\
  selected表示菜单最后遍历，这是默认规则，一般不需要改变"
sortByAttribute:
  - "depth"
  - "list"
  - "selected"
findByDescription: "默认生成控件唯一定位符的表达式风格，可选项 default|android|id|xpath，默认会自动判断是否使用android定\
  位或者ios定位"
findBy: "xpath"
baseUrlDescription: "设置一个起始点，从这个起始点开始计算深度，比如默认从登录后的界面开始计算"
baseUrl: []
appWhiteListDescription: "app白名单，允许在这些app里进行遍历"
appWhiteList: []
urlBlackListDescription: "url黑名单，用于排除某些页面的遍历"
urlBlackList: []
urlWhiteListDescription: "url白名单，仅在这些界面内遍历"
urlWhiteList: []
beforeStartWaitDescription: "启动一个app默认等待的时间"
beforeStartWait: 6000
beforeRestart: []
beforeElementDescription: "在遍历每个控件之前默认执行的动作"
beforeElement: []
afterElementDescription: "在遍历每个控件之后默认执行的动作"
afterElement: []
afterElementWaitDescription: "在遍历每个控件之后默认等待的时间，用于等待新页面加载"
afterElementWait: 500
afterAllDescription: "在遍历完当前页面内的所有控件后，是否需要刷新或者滑动"
afterAll: []
afterAllMaxDescription: "afterAll的最大重试次数，比如连续滑动2次都没新元素即取消"
afterAllMax: 2
tagLimitMaxDescription: "相似控件最多点击几次"
tagLimitMax: 2
tagLimitDescription: "设置部分相似控件的最大遍历次数"
tagLimit:
  - given: []
    when: null
    then: []
    xpath: "确定"
    action: ""
    actions: []
    times: 1000
  - given: []
    when: null
    then: []
    xpath: "取消"
    action: ""
    actions: []
    times: 1000
  - given: []
    when: null
    then: []
    xpath: "share_comment_guide_btn_name"
    action: ""
    actions: []
    times: 1000
assertGlobalDescription: "全局断言"
assertGlobal: []
suiteNameDescription: "报告中的测试套件名字可以由列表内的控件内容替换，增强报告中关键界面的辨识度"
suiteName:
  - "//*[@selected='true']//android.widget.TextView/@text"
screenshotDescription: "是否截图"
screenshot: true
reportTitleDescription: "报告的title"
reportTitle: "AppCrawler"
resultDirDescription: "结果目录，如果为空会自动创建对应时间戳_报名的结果目录"
resultDir: ""
showCancelDescription: "是否展示跳过的控件记录"
showCancel: true
pluginListDescription: "插件列表，暂时禁用，太高级了，很多人不会用"
Description: "。在selectedList firstList lastList等很多配置中，需要填充的是测试步骤Step类型。Step类型由given（\
  满足条件）when（条件满足的行为）then（断言）三部分组成。Step可以简化为xpath（定位表达式，支持xpath 正则 包含关系）与action（点击\
  \ 输入等行为）。"
pluginList: []
```

## 金牌赞助商（Gold Sponsor）

![logo霍格沃兹测试学院](https://ceshiren.com/uploads/default/original/3X/a/2/a270cdc0bf8cb41110ce5999d5278515a0e7cf0f.png)

[霍格沃兹测试开发学社](https://ceshiren.com/)
是 [测吧（北京）科技有限公司](http://qrcode.testing-studio.com/f?from=appcrawler&url=https://ceshiren.com/t/topic/14814)
旗下业界领先的测试开发技术高端教育品牌。 学院课程均由名企一线测试大牛设计，提供实战驱动的系列课程。涵盖移动app自动化测试、接口自动化测试、性能测试、持续集成/持续交付/DevOps 、测试左移、测试右移、测试管理等课程。
[点击学习!](http://qrcode.testing-studio.com/f?from=appcrawler&url=https://testerh.ke.qq.com?flowToken=1040391)

[测吧（北京）科技有限公司](http://qrcode.testing-studio.com/f?from=appcrawler&url=https://ceshiren.com/t/topic/14814)
是一家服务于测试领域的高科技公司，为企业提供全方位的自动化测试技术支持、测试平台开发定制、测试效能提升等咨询与科研合作服务。 先后服务于华为、工信部、信通院等知名企业与机构。

## 编译

```bash
mvn clean package -DskipTests
```

## 技术交流

由霍格沃兹测试开发学社维护，技术交流与issue提交请移步 https://ceshiren.com/c/opensource/appcrawler/ 交流


CHANGELOG.md
# 3.0.0 [TODO]
- 支持从历史数据中寻找最优点击路径
- 支持游戏app遍历
- 支持wetest和mqc
- 改进appium
- 使用节点树模型
- 优化自动重启逻辑
- 简化配置文件
- 完善断言
- 支持性能
- 支持弱网
- 支持接口数据入报告

# 2.7.0
- 代码重构
- 性能提升

仍然不满意，但是应很多公司的使用要求，先提前放出beta版本


# 2.6.0 [TODO]
- 支持accessibility service
- 支持rpc dubbo json-rpc
- 支持云插件
- 改进appium的Uiautomator2的性能与bug

# 2.5
- 支持android native adb遍历 [done]
- 简化遍历逻辑
- 支持根据属性排序遍历

# 2.4
- 支持java10 [done]
- 支持selenium [done]

# 2.3.1
- 合并了androidCapability iosCapability到capability [done]
- 支持定位加速[完成]
- 更新到java-client 6.0，坑好多[Done]

# 2.2.1
- 合并triggerAction firstList lastList为orderList
- 去掉autoCrawl

# 2.2.0
- 修复自动重启执行afterUrlFinish的bug，增强afterUrlFinish支持条件，支持given when
- 增强assert配置的模型，与testcase保持一致，只设置given then即可
- 修复iOS backButton的bug
- 减少getCurrentActivy api调用，会导致appium不稳定
- 修复swipe，这个api仍然不稳定，moveTo是终点位置
- 增加restartOnError命令


# 2.1.2
- 跟进支持appium 1.7[完成]

# 2.1.0
### bugfix
mark图片异常的问题
### 自动化用例
只是demo. 还有很多细节需要设计的更好.  
支持given when then风格, 也支持简化的xpath action then的简单风格.  
```yaml
#设置这个跳过遍历
autoCrawl: false
#测试用例入口
testcase:
#测试用例名字
  name: demo1
  steps:
  - when:
      xpath: //*
      action: driver.swipe(0.5, 0.8, 0.5, 0.2)
  - when:
      xpath: //*
      action: driver.swipe(0.5, 0.2, 0.5, 0.8)
#简化风格. 没有when
  - xpath: 自选
    action: click
    then:
    - //*[contains(@text, "港股")]
```
所有的xpath的设置都支持如下三种形式
- xpath //*[contains(@resource-id, 'ddd')]
- regex ^确定$ 
- contains关系 取消 确定
# 2.0.0
支持macaca[完成]  
失败重试[完成]  
支持简单的测试用例[完成]  
架构重新设计[完成] 
新老版本对比报告改进[完成]  
# 1.9.0
支持遍历断言[完成]  
支持历史对比断言[完成]  
修正不支持uiautomator2的问题[完成]  
支持yaml自动化测试用例[完成]  
action支持长按[完成]  
重构用例生成方式[完成]  

# 1.8.0
对子菜单的支持, 智能判断是否有子菜单  
支持断点续传机制  
支持自动重启appium机制, 用于防止iOS遍历内存占用太大问题  
分离插件到独立项目  

# 1.7.0
android跳到其他app后自动后退[完成]  
截图复用优化提速 [完成]  
报告增加点击前后的截图 [完成]  
独立的report子命令 [完成]  
配置支持动态指令 [完成]  
配置与老版本不兼容 [重要提醒]  
支持自定义报告title [完成]  

# 1.6.0 [内测]
增加动态插件 [完成]  
支持beforeElementAction的afterElementAction配置 [完成]
修复app的http连接支持 [完成]
支持url白名单 [完成]
支持defineUrl的xpath属性提取 [完成]
未遍历控件用测试用例的cancel状态表示 [完成]
两次back之间的时间间隔设定为不低于4s防止粘连 [完成]

# 1.5.0
配置文件内容变更 此版本不再向下兼容, 推荐使用yaml配置文件
标准的html报告 [完成]
windows下中文编码问题 [完成]
windows下命令行超长问题[完成]
加入yaml配置格式支持并添加注释 [完成]
startupActions支持scala表达式 [完成]

# 1.4.0
元素点击之前开始截图并高亮要点击的控件[完成]
修复freemind文件无法打开的问题[完成]
# 1.3.1
增加最大后退尝试次数
增加跳出app的判断
bugfix:
解决文件名特殊符号问题
修复不同界面漏掉截图的问题

# 1.3.0

# 1.2.2
支持相对路径的apk地址.
android的端口指定不再使用4730而是和ios一样
# 1.2.1
url定义优化, 内容变更改进
支持自动化测试. 添加了兼容性测试的例子
# 1.2.0
兼容appium1.5去掉了不支持的findElementByName方法
对xpath元素查找进行了优化 解析dom结构时生成合适的xpath表达式
# 1.1.4
增加log和tagLimit两个插件
截图时间超过5s自动跳过
增加Android和iOS的log输出
增加scroll方向支持
# 1.1.3
自动判断页面是否变化. 界面变化才截图. 能减少大量的重复截图
导出界面dom结构用于diff分析
# 1.1.2
增加-t参数. 支持最大遍历时间
增加-o参数, 支持设定结果目录
# 1.1.1
增加每个url最大滚动的次数
# 1.1.0
增加思维导图生成
增加插件支持
清理大量的无关文件和测试用例

# 1.0.1
支持基本遍历
支持命令行运行方式


src/main/scala/AppCrawler.scala
import org.apache.log4j.{Level, Logger, BasicConfigurator}
import org.scalatest.ConfigMap

/**
  * Created by seveniruby on 16/1/7.
  */


import java.io.File

object AppCrawler extends CommonLog{
  var logPath=""
  case class Param(
                           app: File = new File("."),
                           conf: File = new File("."),
                           verbose: Boolean = false,
                           mode:String="",
                           sbt_params: Seq[String]=Seq(),
                           platform: String = "android",
                           appium:String = "http://127.0.0.1:4723/wd/hub/",
                           resultDir: String = "",
                           maxTime:Int = 3600*3,
                           capability: Map[String, String] = Map[String, String]()
                         )
    def sbt(args: Seq[String]): Unit = {
      import scala.sys.process._
      //val sbt="/usr/local/Cellar/sbt/0.13.11/libexec/sbt-launch.jar"
      val project_dir=getClass.getProtectionDomain.getCodeSource.getLocation.toURI.getPath.
        split("/").dropRight(2).mkString("/")
      val launcherJar = s"${project_dir}/lib/sbt-launch.jar"
      val cmd = Seq("java", "-jar", launcherJar)++args // You
      log.trace(cmd)
      cmd ! ProcessLogger(stdout append _ + "\n", stderr append _ + "\n")
    }


  def main(args: Array[String]) {
    val parser = new scopt.OptionParser[Param]("appcrawler") {
      head(
        """
          |AppCrawler 1.2.0
          |app爬虫, 用于自动遍历测试. 支持Android和iOS, 支持真机和模拟器
          |灵感来源: 晓光 泉龙 杨榕 雪球测试团队出品
          |移动测试技术交流 https://testerhome.com
        """.stripMargin)
      opt[File]('a', "app") action { (x, c) =>{
        c.copy(app = x)
      }
      } text ("Android或者iOS的文件地址, 可以是网络地址, 赋值给appium的app选项")
      opt[File]('c', "conf") action { (x, c) =>
        c.copy(conf = x)
      } validate { x => {
        if (x.isFile) {
          success
        } else {
          failure(s"$x not exist")
        }
      }
      } text ("配置文件地址")
      opt[String]('p', "platform") action { (x, c) =>
        c.copy(platform = x)
      } text ("平台类型android或者ios")
      opt[Int]('t', "maxTime") action { (x, c) =>
        c.copy(maxTime = x)
      } text ("最大运行时间. 单位为秒. 超过此值会退出. 默认最长运行3个小时")
      opt[String]('u', "appium") action { (x, c) =>
        c.copy(appium = x)
      } text ("appium的url地址")
      opt[String]('o', "output") action { (x, c) =>
        c.copy(resultDir = x)
      } text ("遍历结果的保存目录. 里面会存放遍历生成的截图, 思维导图和日志")
      opt[Map[String, String]]("capability") valueName ("k1=v1,k2=v2...") action { (x, c) =>
        c.copy(capability = x)
      } text ("appium capability选项, 这个参数会覆盖-c指定的配置模板参数, 用于在模板配置之上的参数微调")
      opt[Unit]('v', "verbose") action { (_, c) =>
        c.copy(verbose = true)
      } text ("是否展示更多debug信息")
      help("help") text (
        """
          |示例
          |appcrawler -a xueqiu.apk
          |appcrawler -a xueqiu.apk --capability noReset=true
          |appcrawler -c conf/xueqiu.json
          |appcrawler -c xueqiu.json  -p ios --capability udid=[你的udid] -a Snowball.app
          |appcrawler -c xueqiu.json  -p ios -a Snowball.app -u http://127.0.0.1:4730/wd/hub
        """.stripMargin)
      cmd("sbt") action { (_, c) => {
        c.copy(mode = "sbt")
      } } text("sbt是一个调用sbt命令运行测试的开关. 可以传递sbt的参数\n") children(
        arg[String]("<sbt params>...") unbounded() optional() action { (x, c) =>
          c.copy(sbt_params = c.sbt_params :+ x) } text("sbt的参数列表")
        )

    }
    // parser.parse returns Option[C]

    val args_new = if (args.length == 0) {
      Array("--help")
    } else {
      args
    }
    parser.parse(args_new, Param()) match {
      case Some(config) => {
        if(config.verbose){
          logLevel=Level.TRACE
        }
        log.trace("config=")
        log.trace(config)
        if(config.sbt_params.nonEmpty){
          sbt(config.sbt_params)
          return()
        }
        var crawlerConf = new CrawlerConf
        //获取配置模板文件
        if (config.conf.isFile) {
          log.info(s"Find Conf ${config.conf.getAbsolutePath}")
          crawlerConf=crawlerConf.load(config.conf)
        }
        //判断平台
        crawlerConf.currentDriver = config.platform


        val fileName=config.app.getName
        if(fileName.matches(".*\\.apk$")){
          log.info("Set Platform=Android")
          crawlerConf.currentDriver = "Android"
        }
        if(fileName.matches(".*\\.ipa$") || fileName.matches(".*\\.app$") ){
          log.info("Set Platform=iOS")
          crawlerConf.currentDriver = "iOS"
        }

        //合并capability, 命令行>特定平台的capability>通用capability
        crawlerConf.currentDriver.toLowerCase match {
          case "android"=> {
            crawlerConf.capability=crawlerConf.capability++crawlerConf.androidCapability
          }
          case "ios" => {
            crawlerConf.capability=crawlerConf.capability++crawlerConf.iosCapability
          }
        }
        crawlerConf.capability ++= config.capability

        if(config.app.getName.nonEmpty) {
          if(config.app.getPath.matches(".*:/.*")){
            crawlerConf.capability ++= Map("app" -> config.app.getPath.replace(":/", "://"))
          }else{
            crawlerConf.capability ++= Map("app" -> config.app.getAbsolutePath)
          }
          log.info(s"app path = ${crawlerConf.capability("app")}")
        }
        if(config.appium.matches("[0-9]+")){
          crawlerConf.capability++=Map("appium" -> s"http://127.0.0.1:${config.appium}/wd/hub")
        }else{
          crawlerConf.capability++=Map("appium" -> config.appium)
        }
        log.info(s"appium address = ${crawlerConf.capability.get("appium")}")

        crawlerConf.maxTime=config.maxTime
        crawlerConf.resultDir=config.resultDir

        //获得app设置
        //log.trace(s"app path=${config.app.getPath} ${config.app.getName} ${config.app.getAbsolutePath} ${config.app.getCanonicalPath}")
        log.trace(crawlerConf.toJson)
        new AppCrawlerTestCase().execute(configMap = ConfigMap("conf" -> crawlerConf), fullstacks = true)
      }
      case None => {}
    }
  }
}
